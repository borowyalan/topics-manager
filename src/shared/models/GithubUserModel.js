import { UserModel } from './UserModel';
export class GithubUserModel extends UserModel {
    static fromOAuth0(idTokenPayload) {
        return Object.assign(new GithubUserModel(), idTokenPayload);
    }
}
