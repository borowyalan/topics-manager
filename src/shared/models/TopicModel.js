import { UserModel } from './UserModel';
import { find } from 'lodash';

export default class TopicModel {
    title;
    description;
    likes;
    trainers;
    members;
    author; //GithubUserModel
    id;

    static fromBackendData(backendData) {
        const topic = Object.assign(new TopicModel(), backendData);
        topic.members = backendData.members
            ? backendData.members.map(member =>
                Object.assign(new UserModel(), member)
            )
            : [];
        topic.trainers = backendData.trainers
            ? backendData.trainers.map(trainer =>
                Object.assign(new UserModel(), trainer)
            )
            : [];
        return topic;
    }

    amIAttending(myself) {
        return myself && find(this.members, { email: myself.email });
    }

    amITrainer(myself) {
        return (
            myself &&
            (this.author.email === myself.email ||
                find(this.trainers, { email: myself.email }))
        );
    }
}
