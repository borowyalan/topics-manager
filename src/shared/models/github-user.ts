import { User } from './user';

export interface GithubUser extends User {
    idTokenPayload: string;
}
