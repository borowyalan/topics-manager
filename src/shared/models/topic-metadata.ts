export interface TopicMetadata {
    title: string;
    description: string;
}
