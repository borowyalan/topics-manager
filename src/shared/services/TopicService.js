import TopicModel from '../models/TopicModel';

export default class TopicService {

    //TODO: add tests when backend is "added" (called paths)
    //temp
    static people = [
        {email: 'john@doe.com', name: 'John Doe 1'},
        {email: 'john_alpha@doe.com', name: 'John Doe 2'},
        {email: 'john_omega@doe.com', name: 'John Doe 3'}
    ];

    static topics = [
        TopicModel.fromBackendData(
            {
                title: 'React Basic',
                description: 'Lorem ipsum',
                trainers: TopicService.people,
                members: TopicService.people,
                likes: 10,
            }),
        TopicModel.fromBackendData({
            title: 'React Native Basic',
            description: 'Lorem ipsum',
            trainers: TopicService.people,
            members: TopicService.people,
            likes: 39,
        }),
        TopicModel.fromBackendData({
            title: 'React + Redux',
            description: 'Lorem ipsum',
            trainers: TopicService.people,
            members: TopicService.people,
            likes: 13,
        }),
        TopicModel.fromBackendData({
            title: 'React + Redux + Thunk',
            description: 'Lorem ipsum',
            trainers: TopicService.people,
            members: TopicService.people,
            likes: 0,
        }),
    ];

    static async getSubmittedTopics() {
        return this.topics;
    }

    static async getPastTopics() {
        return this.topics;
    }

    static submitTopic(topic, author) {
        // eslint-disable-next-line no-undef
        const db = firebase.database();
        const newTopicKey = db.ref().child('topics').push().key;
        //temp; uncomment to push something to realtime database
        db.ref('topics/' + newTopicKey).set({
            ...topic, author: author.strip(), trainers: [], users: []
        });
    }
}
