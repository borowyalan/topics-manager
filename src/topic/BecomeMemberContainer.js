import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TopicModel from '../shared/models/TopicModel';
import { becomeMember, leaveTopic } from '../actions/member.actions';
import 'react-awesome-popover/dest/react-awesome-popover.css';
import JoinWorkshopContainer from './components/JoinWorkshopContainer';
import { UserModel } from '../shared/models/UserModel';

class BecomeMemberContainer extends React.Component {
    leave = () => {
        const { topic, user, leave } = this.props;
        leave(topic, user);
    };

    attend = () => {
        const { topic, user, attend } = this.props;
        attend(topic, user);
    };

    render() {
        const { topic, user } = this.props;
        return (
            <JoinWorkshopContainer
                alreadyAttending={topic.amIAttending(user)}
                anonymous={!user}
                joinText="Chcę być uczestnikiem"
                leaveText="Rezygnuję"
                attend={this.attend}
                leave={this.leave}
            />
        );
    }
}

const mapStateToProps = state => ({
    user: state.auth.user,
});

const mapDispatchToProps = dispatch => ({
    attend: (topic, user) => dispatch(becomeMember(topic, user)),
    leave: (topic, user) => dispatch(leaveTopic(topic, user)),
});

BecomeMemberContainer.propTypes = {
    attend: PropTypes.func.isRequired,
    leave: PropTypes.func.isRequired,
    user: PropTypes.instanceOf(UserModel),
    topic: PropTypes.instanceOf(TopicModel).isRequired,
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(BecomeMemberContainer);
