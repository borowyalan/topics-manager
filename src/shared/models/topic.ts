import { find } from 'lodash';
import { TopicMetadata } from './topic-metadata';
import { User } from './user';

export interface Topic extends TopicMetadata {
    trainers: User[];
    members: User[];
    author: User;
    id: string;
}

export const isAttending = (topic: Topic, user: User) => {
    return Boolean(user && find(topic.members, { email: user.email }));
};

export const isTrainer = (topic: Topic, user: User) => {
    return Boolean(
        user &&
            (topic.author.email === user.email ||
                find(topic.trainers, { email: user.email }))
    );
};
