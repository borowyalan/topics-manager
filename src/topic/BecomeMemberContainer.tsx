import React from 'react';
import 'react-awesome-popover/dest/react-awesome-popover.css';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { becomeMember, leaveTopic } from '../actions/member.actions';
import { isAttending, Topic } from '../shared/models/topic';
import { User } from '../shared/models/user';
import { JoinWorkshopContainer } from './components/JoinWorkshopContainer';

interface OwnProps {
    topic: Topic
}

interface StateProps {
    user: User
}

interface DispatchProps {
    attend: (topic: Topic, user: User) => void,
    leave: (topic: Topic, user: User) => void,
}

type Props = OwnProps & StateProps & DispatchProps

class BecomeMemberContainer extends React.Component<Props> {

    public render() {
        const { topic, user } = this.props;
        return (
            <JoinWorkshopContainer
                alreadyAttending={isAttending(topic, user)}
                anonymous={!user}
                joinText="Chcę być uczestnikiem"
                leaveText="Rezygnuję"
                attend={this.attend}
                leave={this.leave}
            />
        );
    }

    private leave = () => {
        const { topic, user, leave } = this.props;
        leave(topic, user);
    };

    private attend = () => {
        const { topic, user, attend } = this.props;
        attend(topic, user);
    };

}

const mapStateToProps = state => ({
    user: state.auth.user,
});

const mapDispatchToProps = (dispatch: Dispatch<any>): DispatchProps => ({
    attend: (topic, user) => dispatch(becomeMember(topic, user)),
    leave: (topic, user) => dispatch(leaveTopic(topic, user)),
});

export default connect<StateProps, DispatchProps, OwnProps>(
    mapStateToProps,
    mapDispatchToProps,
)(BecomeMemberContainer);
