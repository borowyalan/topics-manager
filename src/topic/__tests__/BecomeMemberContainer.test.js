import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureMockStore from 'redux-mock-store';
import 'jest-styled-components';
import thunk from 'redux-thunk';
import TopicModel from '../../shared/models/TopicModel';
import { GithubUserModel } from '../../shared/models/GithubUserModel';
import BecomeMemberContainer from '../BecomeMemberContainer';
import TopicService from '../../shared/services/TopicService';
import {
    MEMBER_LEAVE_REQUEST,
    MEMBER_SUBMIT_REQUEST,
} from '../../actions/action_types';

const mockedActionType = 'TopicContainerAction';

jest.mock('../../actions/topic.actions', () => ({
    getTopics: () => ({ type: mockedActionType }),
}));

configure({ adapter: new Adapter() });
const mockStore = configureMockStore([thunk]);

describe('<BecomeMemberContainer>', () => {
    const user = GithubUserModel.fromBackend({
        email: 'john@doe.com',
    });

    const topicAttendingTo = TopicModel.fromBackendData({
        members: [user],
    });

    const topicNotAttendingTo = TopicModel.fromBackendData({
        members: [],
    });

    const initialState = {
        auth: {
            user,
        },
    };

    let wrapper;
    let store = mockStore(initialState);

    describe('when not attending to given topic yet', () => {
        let fakeService;
        beforeEach(() => {
            store.clearActions();
            fakeService = jest
                .spyOn(TopicService, 'attend')
                .mockImplementation(() => Promise.resolve());
            wrapper = shallow(
                <BecomeMemberContainer
                    store={store}
                    topic={topicNotAttendingTo}
                />
            ).dive();
        });

        afterEach(() => {
            fakeService.mockRestore();
        });

        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should have user connected correctly from redux', () => {
            expect(wrapper.instance().props.user).toBe(initialState.auth.user);
        });

        describe('when button is pressed', () => {
            beforeEach(() => {
                wrapper
                    .find('BecomeMemberContainer__ActionButton')
                    .at(0)
                    .props()
                    .onClick();
            });

            it('should dispatch MEMBER_SUBMIT_REQUEST action', () => {
                expect(store.getActions()[0].type).toBe(MEMBER_SUBMIT_REQUEST);
            });
        });
    });

    describe('when already attending to given topic', () => {
        let fakeService;
        beforeEach(() => {
            store.clearActions();
            fakeService = jest
                .spyOn(TopicService, 'leave')
                .mockImplementation(() => Promise.resolve());
            wrapper = shallow(
                <BecomeMemberContainer store={store} topic={topicAttendingTo} />
            ).dive();
        });

        afterEach(() => {
            fakeService.mockRestore();
        });

        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('should have user connected correctly from redux', () => {
            expect(wrapper.instance().props.user).toBe(initialState.auth.user);
        });

        describe('when button is pressed', () => {
            beforeEach(() => {
                wrapper
                    .find('BecomeMemberContainer__ActionButton')
                    .at(0)
                    .props()
                    .onClick();
            });

            it('should dispatch MEMBER_LEAVE_REQUEST action', () => {
                expect(store.getActions()[0].type).toBe(MEMBER_LEAVE_REQUEST);
            });
        });
    });
});
