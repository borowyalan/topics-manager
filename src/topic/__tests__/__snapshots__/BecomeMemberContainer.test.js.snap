// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BecomeMemberContainer> when already attending to given topic should match snapshot 1`] = `
<div>
  <BecomeMemberContainer__ActionButton
    onClick={[Function]}
  >
    <BecomeMemberContainer__ActionButtonSpanText>
      Rezygnuję
    </BecomeMemberContainer__ActionButtonSpanText>
  </BecomeMemberContainer__ActionButton>
</div>
`;

exports[`<BecomeMemberContainer> when not attending to given topic yet should match snapshot 1`] = `
<div>
  <BecomeMemberContainer__ActionButton
    onClick={[Function]}
  >
    <BecomeMemberContainer__ActionButtonSpanText>
      Chcę być uczestnikiem
    </BecomeMemberContainer__ActionButtonSpanText>
  </BecomeMemberContainer__ActionButton>
</div>
`;
